771 worst possible result

franchuterivera opened the pull request:
Created a function that based on the metric determines the worst cost for crash

Smac uses a MAX int defined as a custom constant, I am just picking a value given by NumPy which is already loaded

 I updated a test to make sure metrics whose optimum is zero are taken into account for coverage purposes.

mfeurer commented on autosklearn/evaluation/__init__.py:
```diff
@@ -35,13 +33,28 @@ def fit_predict_try_except_decorator(ta, queue, **kwargs):
+    # assume 1 as the worst cost as autosklearn works with err
+    worst_possible_result = 1.0
+
+    if metric is not None:
```
Hey, how can the metric be `None` here?

mfeurer commented on autosklearn/evaluation/__init__.py:
```diff
@@ -35,13 +33,28 @@ def fit_predict_try_except_decorator(ta, queue, **kwargs):
+        # Else, we use the make_scorer info to define the
+        # worst possible result
+        if metric._sign < 0:
+            worst_possible_result = np.iinfo(np.uint32).max
```
As there every metric has the attribute `optimum`, I'm actually wondering whether we should add a new attribute `worst_possible_value` to each metric so that it's clearly defined (and only in one place, namely the metrics module) how they work. What do you think?

franchuterivera commented:
- Make sure that a metric that inherits from scorer is passed (just a paranoia and in-code test)
- Now the metric could be none, when doing unit testing. I updated the automl class to handle a metric provided by the user. In the normal workflow, the estimators make sure that if the metric is not, accuracy for instance is used.
- added a comment to clarify  _loss functionality.
- The expected workflow:

1. A metric has optimum and worst results.
2. As autosklearn minimizes, a function get_worst_posible result translates the metric._worst_possible_result in a way the a minimization engine would understand that this is a worst possible result. Example: 
For accuracy, optimum=1 and worst=0. The worst possible error is optimum-worst=1-0=1. For
For logloss= optimum=0 and worst=MAXTINT= worst possible error is MAXINT

mfeurer commented on autosklearn/automl.py:
```diff
@@ -89,6 +89,7 @@ def __init__(self,
                  get_smac_object_callback=None,
                  smac_scenario_args=None,
                  logging_config=None,
+                 metric=None,
```
I don't quite see why you pass the metric in here and do not use it later on.

franchuterivera replied:
To have it documented here:
We are adding the metric in the class instantiation because the automl object gets created directly during unit testing so that the dummy classifier can be tested.

Not needed during the normal auto-sklearn flow. Long term we would cleanup the fit/related methods because currently both the fit and instantiation takes the metric as argument.

mfeurer commented on test/test_metric/test_metrics.py:
```diff
@@ -116,7 +118,7 @@ def test_proba_scorer_binary(self):
 
         scorer = autosklearn.metrics._ProbaScorer(
-            'accuracy', sklearn.metrics.log_loss, 0, 1, {})
+            'accuracy', sklearn.metrics.log_loss, 0, MAXINT, 1, {})
```
Actually, as you're touching these lines, could you please change the names of the scorers?

mfeurer commented:
Closes #771
